source/Sagan-RDBMS/RDMBSRepository.class.st

```diff
@@ -0,0 +1,214 @@
+Class {
+	#name : #RDMBSRepository,
+	#superclass : #Repository,
+	#instVars : [
+		'descriptorSystem',
+		'modelObjectClass',
+		'conflictCheckingStrategy',
+		'sessionProvider'
+	],
+	#category : #'Sagan-RDBMS-Core'
+}
+
+{ #category : #'instance creation' }
+RDMBSRepository class >> storingObjectsOfType: aDomainObjectClass checkingConflictsAccordingTo: aConflictCheckingStrategy workingWith: aSessionProvider [
+
+	^ self new
+		initializeStoringObjectsOfType: aDomainObjectClass
+		checkingConflictsAccordingTo: aConflictCheckingStrategy
+		workingWith: aSessionProvider
+]
+
+{ #category : #mappings }
+RDMBSRepository >> beAwareOfClassModelDefinedBy: aClassModelDefinition [
+
+	self
+		withSessionDo: [ :session | session system addClassModelDefinedBy: aClassModelDefinition ]
+]
+
+{ #category : #mappings }
+RDMBSRepository >> beAwareOfDescriptorDefinedBy: aDescriptorDefinition [
+
+	self
+		withSessionDo: [ :session | session system addDescriptorDefinedBy: aDescriptorDefinition ]
+]
+
+{ #category : #mappings }
+RDMBSRepository >> beAwareOfTableDefinedBy: aTableDefinition [
+
+	self withSessionDo: [ :session | session system addTableDefinedBy: aTableDefinition ]
+]
+
+{ #category : #mappings }
+RDMBSRepository >> beAwareOfTypeResolverDefinedBy: aTypeResolverDefinition [
+
+	self
+		withSessionDo: [ :session | session system addTypeResolverDefinedBy: aTypeResolverDefinition ]
+]
+
+{ #category : #mappings }
+RDMBSRepository >> configureMappingsIn: aBlock [
+
+	aBlock cull: self.
+	self withSessionDo: [ :session | session system: session system ]
+]
+
+{ #category : #'private - accessing' }
+RDMBSRepository >> conflictCheckingStrategy [
+
+	^ conflictCheckingStrategy
+]
+
+{ #category : #querying }
+RDMBSRepository >> countMatching: aCriteriaOrBlock [
+
+	^ self
+		withSessionDo: [ :session | self handleReadErrorsDuring: [ session count: modelObjectClass where: ( self asMatchingCriteria: aCriteriaOrBlock ) ] ]
+]
+
+{ #category : #'private - accessing' }
+RDMBSRepository >> executeQuery: query [
+
+	^ self withSessionDo: [ :session | self handleReadErrorsDuring: [ session execute: query ] ]
+]
+
+{ #category : #querying }
+RDMBSRepository >> findAll [
```

Comment:
Uses of `withSessionDo: ... handleReadErrorsDuring:` should be changed to use a new method `withSessionRead:`or `inSessionRead:` and the same for writing.